src/envs/blocks.py

```diff
@@ -0,0 +1,439 @@
+                 if b.type == self._block_type]
+        existing_xys = {(float(p[0]), float(p[1])) for p in poses}
+        if not self._table_xy_is_clear(x, y, existing_xys):
+            return next_state
+        # Execute putontable
+        next_state.set(block, "pose_x", x)
+        next_state.set(block, "pose_y", y)
+        next_state.set(block, "pose_z", z)
+        next_state.set(block, "held", 0.0)
+        next_state.set(block, "clear", 1.0)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def _transition_stack(self, state: State, action: Action) -> State:
+        next_state = state.copy()
+        # Can only stack if fingers are closed
+        if state.get(self._robot, "fingers") >= self.open_fingers:
+            return next_state
+        # Check that both blocks exist
+        block = self._get_held_block(state)
+        assert block is not None
+        x, y, z, fingers = action.arr
+        other_block = self._get_highest_block_below(state, x, y, z)
+        if other_block is None:  # no block to stack onto
+            return next_state
+        # Can't stack onto yourself!
+        if block == other_block:
+            return next_state
+        # Need block we're stacking onto to be clear
+        if state.get(other_block, "clear") < self.clear_tol:
+            return next_state
+        # Execute stack by snapping into place
+        cur_x = state.get(other_block, "pose_x")
+        cur_y = state.get(other_block, "pose_y")
+        cur_z = state.get(other_block, "pose_z")
+        next_state.set(block, "pose_x", cur_x)
+        next_state.set(block, "pose_y", cur_y)
+        next_state.set(block, "pose_z", cur_z+CFG.blocks_block_size)
+        next_state.set(block, "held", 0.0)
+        next_state.set(block, "clear", 1.0)
+        next_state.set(other_block, "clear", 0.0)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def get_train_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_train_tasks,
+                               possible_num_blocks=CFG.blocks_num_blocks_train,
+                               rng=self._train_rng)
+
+    def get_test_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_test_tasks,
+                               possible_num_blocks=CFG.blocks_num_blocks_test,
+                               rng=self._test_rng)
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {self._On, self._OnTable, self._GripperOpen, self._Holding,
+                self._Clear}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._block_type, self._robot_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {self._Pick, self._Stack, self._PutOnTable}
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [x, y, z, fingers]
+        return Box(-100.0, 100.0, (4,))
```

Comment:
any reason why these bounds are so big? the experience with -1000 in the covers environment leading to issues with predicate learning for amber made me weary of large magic numbers

Reply:
fixed